cohort-evaluator-spark/src/main/java/com/ibm/cohort/cql/spark/data/SparkSchemaCreator.java

```diff
@@ -0,0 +1,202 @@
+		this.requests = requests;
+		this.contextDefinitions = contextDefinitions;
+	}
+
+	public Map<String, StructType> calculateSchemasForContexts(List<String> contextNames) throws Exception {
+		HashMap<String, StructType> contextResultSchemas = new HashMap<>();
+
+		for (String contextName : contextNames) {
+			StructType schema = calculateSchemaForContext(contextName);
+			if (schema != null) {
+				contextResultSchemas.put(contextName, schema);
+			}
+		}
+		
+		return contextResultSchemas;
+	}
+
+	public StructType calculateSchemaForContext(String contextName) throws Exception {
+		List<CqlEvaluationRequest> filteredRequests = requests.getEvaluations().stream().filter(r -> r.getContextKey().equals(contextName)).collect(Collectors.toList());
+
+		StructType resultsSchema = new StructType();
+
+		Set<Tuple2<String, String>> usingInfo = new HashSet<>();
+
+		for (CqlEvaluationRequest filteredRequest : filteredRequests) {
+			CqlLibraryDescriptor descriptor = filteredRequest.getDescriptor();
+			String measureName = descriptor.getLibraryId();
+
+			for (String expression : filteredRequest.getExpressions()) {
+				Library library = CqlLibraryReader.read(
+						libraryProvider.getLibrary(new CqlLibraryDescriptor().setLibraryId(measureName).setVersion(descriptor.getVersion())).getContentAsStream()
+				);
+
+				// Track the set of non-system using statements across measures.
+				// Information is used later to access ModelInfos when searching
+				// for context key column type information.
+				usingInfo.addAll(library.getUsings().getDef().stream()
+										 .filter(x -> !x.getLocalIdentifier().equals("System"))
+										 .map(x -> new Tuple2<>(x.getLocalIdentifier(), x.getVersion()))
+										 .collect(Collectors.toList()));
+
+				List<ExpressionDef> expressionDefs = library.getStatements().getDef().stream()
+						.filter(x -> x.getName().equals(expression))
+						.collect(Collectors.toList());
+
+				if (expressionDefs.isEmpty()) {
+					throw new IllegalArgumentException("Expression " + expression + " is configured in the CQL jobs file, but not found in "
+															   + descriptor.getLibraryId() + "." + descriptor.getVersion());
+				}
+
+				QName resultTypeName = expressionDefs.get(0).getExpression().getResultTypeName();
+				resultsSchema = resultsSchema.add(measureName + "." + expression, QNameToDataTypeConverter.getFieldType(resultTypeName), true);
+			}
+		}
+
+		if (resultsSchema.fields().length > 0) {
+			Tuple2<String, DataType> keyInformation = getKeyInformationForContext(contextName, usingInfo);
+			StructType fullSchema = new StructType()
+					.add(keyInformation._1(), keyInformation._2(), false);
+
+			for (StructField field : resultsSchema.fields()) {
+				fullSchema = fullSchema.add(field);
+			}
+			resultsSchema = fullSchema;
+		}
+		
+		return resultsSchema;
+	}
+
+	private Tuple2<String, DataType> getKeyInformationForContext(String contextName, Set<Tuple2<String, String>> usingInfos) {
```

Comment:
Should this be called `getDataTypeForContextKey`?

Reply:
Sure